cmd/seed-controller-manager/controllers.go

```diff
@@ -84,26 +83,36 @@ func createSeedConditionUpToDateController(ctrlCtx *controllerContext) error {
 }
 
 func createClusterComponentDefaulter(ctrlCtx *controllerContext) error {
-	defaultCompontentsOverrides := kubermaticv1.ComponentSettings{
-		Apiserver: kubermaticv1.APIServerSettings{
-			DeploymentSettings:          kubermaticv1.DeploymentSettings{Replicas: utilpointer.Int32Ptr(int32(ctrlCtx.runOptions.apiServerDefaultReplicas))},
-			EndpointReconcilingDisabled: utilpointer.BoolPtr(ctrlCtx.runOptions.apiServerEndpointReconcilingDisabled),
-		},
-		ControllerManager: kubermaticv1.DeploymentSettings{
-			Replicas: utilpointer.Int32Ptr(int32(ctrlCtx.runOptions.controllerManagerDefaultReplicas))},
-		Scheduler: kubermaticv1.DeploymentSettings{
-			Replicas: utilpointer.Int32Ptr(int32(ctrlCtx.runOptions.schedulerDefaultReplicas))},
+	defaultOverrides, err := readDefaultComponentsSettings(ctrlCtx)
+	if err != nil {
+		return err
 	}
+
 	return clustercomponentdefaulter.Add(
 		context.Background(),
 		ctrlCtx.log,
 		ctrlCtx.mgr,
 		ctrlCtx.runOptions.workerCount,
-		defaultCompontentsOverrides,
+		defaultOverrides,
 		ctrlCtx.runOptions.workerName,
 	)
 }
 
+func readDefaultComponentsSettings(ctrlCtx *controllerContext) (kubermaticv1.ComponentSettings, error) {
+	ret := kubermaticv1.ComponentSettings{}
+	cm := &corev1.ConfigMap{}
+	r := ctrlCtx.mgr.GetAPIReader()
+	ns := ctrlCtx.runOptions.namespace
+	n := ctrlCtx.runOptions.clusterComponentsSettings
+	if err := r.Get(ctrlCtx.ctx, types.NamespacedName{Namespace: ns, Name: n}, cm); err != nil {
+		return ret, fmt.Errorf("get components settings configmap: %w", err)
+	}
```

Comment:
I'm not very fond of this way of consuming a ConfigMap, I would rather mount it as a volume to make explicit the dependency in the manifests (also avoiding potential transient ordering issues) and simply read the file.

Reply:
After thinking a bit I think that turning `ComponentSettings` into a CRD with a controller taking care of it could be even better option. I think we could do something similar to: https://github.com/openshift/cluster-version-operator/blob/2114f132b6a1e9149d864019162fb795da8d4e3d/pkg/cvo/cvo.go#L545

We would gain:
- Better validation
- Easy way to take changes to the default values dynamically

@xrstf WDYT?

Reply:
I like the idea of passing configmap name and mounting it. I like flexibility of it and that it makes it explicit that config is not dynamic.
I don't like crd for component overrides because I think we will ever have 1 crd instance at a time and having multiple component override crd instances is no use.

Reply:
> I like the idea of passing configmap name and mounting it. I like flexibility of it and that it makes it explicit that config is not dynamic.

We had a discussion today during our daily and I think that having config changes taken into account dynamically and benefitting from validation, defaulting etc. would not be bad. On that respect the CRD approach would make sense. We discussed two possible approaches that could be considered:

- Having the component settings in Seed resource: https://github.com/kubermatic/kubermatic/pull/5821/files
- Having a dedicated CRD: https://github.com/kubermatic/kubermatic/pull/5822/files

IMO One important thing to take into account is that whatever approach we take, we have to make sure not to impact existing clusters if defaults are changed. I think the ideal would be to have the defaulting of Clusters `componentsOverrides` done at admission phase and not in a controller, because it should be edge-driven and not level-driven. We should basically do a snapshot of the default component settings at the moment of the Cluster creation (see https://github.com/irozzo-1A/kubermatic-1/blob/2f04e66bc4c354ca7b3f10bc82099347344472e6/pkg/controller/seed-controller-manager/clustercomponentdefaulter/clustercomponentdefaulter.go#L132-L162). 

> I don't like crd for component overrides because I think we will ever have 1 crd instance at a time and having multiple component override crd instances is no use.

I see your point and I think it's valid, but on the other hand we could think about extending this mechanism in the future and having a sort of component setting presets or classes, that can be referenced by user clusters instead of configuring each of them individually. Do you think this would be useful for you?

@xrstf @moelsayed @furkhat Thoughts?

Reply:
For a long run it seems custom CRD would be better. I will start to work on it if you don't have additional comments here @xrstf @moelsayed.

Reply:
I think extending our Seed CRD instead could save us lots of work w.r.t to validation. We already have a webhook and logic to ensure currently in-use settings are not accientally removed (like datacenters).

Reply:
@xrstf I'm not sure it would save lot of work. For sure it would be cheaper, but this should not be our single and ultimate design goal.

> We already have a webhook and logic to ensure currently in-use settings are not accientally removed

I'm not sure this is relevant. What I proposed, I don't know if I have consensus here, is that when clusters get created, we take a snapshot of the referenced component settings. This means that if this get removed there should be no impact at all on the existing `Cluster`s. This would be similar to the relationship between `StorageClass` and `PersistentVolume`.

@furkhat As this is not trivial as it seamed at first glance I would propose to have a proposal first so that we can agree on where we want to go.

Reply:
But then updating the components override will not affected existing clusters, isn't that a major downside? You as the cluster administrator want to ensure a consistent set of overrides and then later you change your mind and now you have to manually edit all existing clusters. That IMHO ruins the usefulness of this feature entirely.

Reply:
I cannot say wether it is a downside or not if we don't first agree on what we want to achieve. What I had in mind is actually classical `defaulting`, what you proposed is more like having two sources of config that get's merged to define the final spec. I think that some specification is definitely needed here.

Reply:
> But then updating the components override will not affected existing clusters, isn't that a major downside?

@xrstf Hm, I haven't thought about default component overrides to be part of continuous reconcile. Specifically for our needs we wanted them to apply for new clusters only during creation and allow independent change of per cluster component override by editing cluster resource `.spec.componentOverrides`.  I will think about it.

> @furkhat As this is not trivial as it seamed at first glance I would propose to have a proposal first so that we can agree on where we want to go.

@irozzo-1A good idea. I will prepare a non formal proposal and put it here.